data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestMapper.java

```diff
@@ -193,14 +217,61 @@ private static String encoding(final DataPlaneContract.ContentMode contentMode)
     };
   }
 
-  private static class Producer<K, V> {
+  private static class ReferenceCounter<T> {
+    final T value;
+    int count;
+
+    public ReferenceCounter(T value) {
+      this.value = value;
+      this.count = 0;
+    }
+
+    public T getValue() {
+      return value;
+    }
+
+    public void increment() {
+      this.count++;
+    }
+
+    /**
+     * @return true if the count is 0, hence nobody is referring anymore to this value
+     */
+    public boolean decrement() {
```

Comment:
I'm trying to find a better name given the return type. :thinking:

Reply:
`decrementAndCheck`?

Reply:
It seems ok to me.